kernel/src/grant.rs

```diff
@@ -72,190 +89,384 @@ impl<'a, T: Default> AppliedGrant<'a, T> {
+        }
+    }
+
+    /// Return an `AppliedGrant` for a grant in a process if the process is
+    /// valid and that grant has already been allocated, or `None` otherwise.
+    fn new_if_allocated(grant: &Grant<T>, process: &'a dyn ProcessType) -> Option<Self> {
+        if let Some(is_allocated) = process.grant_is_allocated(grant.grant_num) {
+            if is_allocated {
+                Some(AppliedGrant {
                     process: process,
                     grant_num: grant.grant_num,
-                    grant: &mut *(typed_grant_pointer as *mut T),
                     _phantom: PhantomData,
                 })
+            } else {
+                // Grant has not been allocated.
+                None
             }
         } else {
-            Err(Error::InactiveApp)
+            // Process is invalid.
+            None
         }
     }
 
-    /// Return an `AppliedGrant` for a grant in a process if that
-    /// grant has already been allocated.
-    ///
-    /// On `Err()`, returns `Err(None)` if the grant has not been allocated
-    /// for this process. Returns `Err(Some(Error))` with an appropriate
-    /// error otherwise.
-    fn get_if_allocated(
-        grant: &Grant<T>,
-        process: &'a dyn ProcessType,
-    ) -> Result<Self, Option<Error>> {
-        process
-            .get_grant_ptr(grant.grant_num)
-            .map_or(Err(None), |grant_ptr| {
-                if grant_ptr.is_null() {
-                    // Grant has not been allocated.
-                    Err(None)
-                } else if grant_ptr == (!0 as *mut u8) {
-                    // A grant pointer set to 0xFFFFFFFF is a flag signaling
-                    // that the grant has already been entered.
-                    Err(Some(Error::AlreadyInUse))
-                } else {
-                    Ok(AppliedGrant {
-                        process: process,
-                        grant_num: grant.grant_num,
-                        grant: unsafe { &mut *(grant_ptr as *mut T) },
-                        _phantom: PhantomData,
-                    })
-                }
-            })
+    /// Return the AppId of the process this AppliedGrant is applied for.
+    pub fn appid(&self) -> AppId {
+        self.process.appid()
     }
 
     /// Run a function with access to the contents of the grant region.
-    /// This is "entering" the grant region, and the _only_ time when
-    /// the contents of a grant region can be accessed.
+    ///
+    /// This is "entering" the grant region, and the _only_ time when the
+    /// contents of a grant region can be accessed.
+    ///
+    /// Note, a grant can only be entered once at a time. Attempting to call
+    /// `.enter()` on a grant while it is already entered will result in a
+    /// panic!()`. See the comment in `access_grant()` for more information.
     pub fn enter<F, R>(self, fun: F) -> R
     where
         F: FnOnce(&mut Borrowed<T>, &mut Allocator) -> R,
```

Comment:
What was the design rationale for passing `Allocator`? It seems like every use case is `|app, _|`.

Reply:
Ah, yeah, that makes sense. I'd suggest `.enter_with_allocator`. If separating the two cases issues I think it's OK to pass the allocator; it would be useful if the documentation talked about it and when/how it is used.